lib/linalg/Matrix.cpp

```diff
@@ -1790,45 +1790,106 @@ const
 void
 Matrix::orthogonalize()
 {
-    for (int work = 1; work < d_num_cols; ++work) {
+    for (int work = 0; work < d_num_cols; ++work)
+    {
+        // Orthogonalize the column.
         double tmp;
-        for (int col = 0; col < work; ++col) {
+        for (int col = 0; col < work; ++col)
+        {
             double factor = 0.0;
             tmp = 0.0;
-            for (int i = 0; i < d_num_rows; ++i) {
-                tmp += item(i, col)*item(i, work);
-            }
-            if (d_num_procs > 1) {
-                MPI_Allreduce(&tmp,
-                              &factor,
-                              1,
-                              MPI_DOUBLE,
-                              MPI_SUM,
+
+            for (int i = 0; i < d_num_rows; ++i)
+                tmp += item(i, col) * item(i, work);
+
+            if (d_distributed && d_num_procs > 1)
+            {
+                MPI_Allreduce(&tmp, &factor, 1, MPI_DOUBLE, MPI_SUM,
                               MPI_COMM_WORLD);
             }
-            else {
+            else
+            {
                 factor = tmp;
             }
-
-            for (int i = 0; i < d_num_rows; ++i) {
-                item(i, work) -= factor*item(i, col);
-            }
+            for (int i = 0; i < d_num_rows; ++i)
+                item(i, work) -= factor * item(i, col);
         }
+
+        // Normalize the column.
         double norm = 0.0;
         tmp = 0.0;
-        for (int i = 0; i < d_num_rows; ++i) {
-            tmp += item(i, work)*item(i, work);
-        }
-        if (d_num_procs > 1) {
+
+        for (int i = 0; i < d_num_rows; ++i)
+            tmp += item(i, work) * item(i, work);
+
+        if (d_distributed && d_num_procs > 1)
+        {
             MPI_Allreduce(&tmp, &norm, 1, MPI_DOUBLE, MPI_SUM, MPI_COMM_WORLD);
         }
-        else {
+        else
+        {
             norm = tmp;
         }
-        norm = sqrt(norm);
-        for (int i = 0; i < d_num_rows; ++i) {
-            item(i, work) /= norm;
+        if (norm > 1.0e-15)
```

Comment:
Should there be a `CAROM_VERIFY(norm > 1.0e-15);`? If the column becomes 0, shouldn't that be a failure?

Reply:
@cval26 are you going to address this?

Reply:
Yes, I'm going to make the required changes this week.

Reply:
I don't think it should be a failure if a column is identically zero. For instance, when using Laghos EQP with the `load` offset option, the first NNLS constraints column can be zero, because the first snapshot vectors are identically zero.

Reply:
Added a private double `zero_tol` variable and a public method for setting its value. Default value is `1e-15`.

Reply:
I don't think it should be a failure if a column becomes zero. This means that the input matrix had linearly dependent columns, so the method has reduced one column to essentially zero.